### PR TITLE
feat(dup): implement pegasus_mutation_duplicator

### PR DESCRIPTION
**Pull request by neverchanje**
### What problem does this PR solve? 

Implement `dsn::replication::mutation_duplicator` interface to support sending mutations to
the remote Pegasus cluster.

Our target is to simplify the "mutation sending" process, reuse the existing client lib, duplicate each
mutation independently, leave optimization to the future. Therefore the performance may be poor,
but the overall cost is acceptable, as long as we make it easy to understand.

[include/dsn/dist/replication/mutation_duplicator.h](https://github.com/XiaoMi/rdsn/blob/master/include/dsn/dist/replication/mutation_duplicator.h)

```cpp
class mutation_duplicator : public replica_base
{
public:
    typedef std::function<void(size_t /*total_shipped_size*/)> callback;

    /// Duplicate the provided mutations to the remote cluster.
    /// The implementation must be non-blocking.
    ///
    /// \param cb: Call it when all the given mutations were sent successfully
    virtual void duplicate(mutation_tuple_set mutations, callback cb) = 0;
};
```

Docs here might be helpful to understand this PR: https://pegasus-kv.github.io/2019/06/09/duplication-design.html#%E6%B5%81%E7%A8%8B

### What is changed and how it works?

The design inside the box is:

1. `pegasus_mutation_duplicator` receives a batch of mutations to be duplicated to the remote endpoint. After sending the entire batch, it calls `cb` for the next batch. The batch-by-batch design
is to preserve the data consistency of both clusters.

```cpp
void duplicate(mutation_tuple_set muts, callback cb) override;
```

2. In our initial implementation, the batch of mutations are shipped one-by-one for consistency:

```cpp
private:
    std::deque<duplicate_rpc> _inflights;
```

We can improve this design with slightly increased complexity, by dividing the mutations into
groups, isolated by their key. This way helps increase concurrency with uncompromising consistency.

```cpp
private:
    // hash -> duplicate_rpc
    std::map<uint64_t, std::deque<duplicate_rpc>> _inflights;
```

3. Types of writes are wrapped in `duplicate_rpc`. This design is extensible for adding and removing a write RPC in the future because the real content is hidden under `task_code` and `raw_message`. The duplicator is unaware of what the write is.

```thrift
struct duplicate_request
{
    // The timestamp of this write.
    1: optional i64 timestamp

    // The code to identify this write.
    2: optional dsn.task_code task_code

    // The binary form of the write.
    3: optional dsn.blob raw_message
}

struct duplicate_response
{
    1: optional i32 error;
}
```

One corner case is when the write type is `RPC_RRDB_RRDB_DUPLICATE`, the duplicator should ignore the write duplicated from others. Because duplication is an one-way edge between two nodes.

If the administrator wants to create a two-way duplication between A and B, it should add_dup on A->B as well as B->A.

### Check List 

Tests 

- Unit test

### CODE REVIEW

File: src/idl/rrdb.thrift [hunk 1]
```diff
@@ -251,6 +251,19 @@ struct scan_response
     6:string        server;
 }
 
+struct duplicate_request
+{
+    1:i64                     timetag;
+    2:dsn.task_code           task_code;
+    3:dsn.blob                raw_message;
+    4:i64                     hash;
```
Comment (hycdong) [inline src/idl/rrdb.thrift]: 这个hash是谁的hash? 从代码中看是rocksdb的key的hash？要不要改成key_hash？

File: src/server/pegasus_mutation_duplicator.cpp [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+                                                         dsn::string_view app)
+    : mutation_duplicator(r), _remote_cluster(remote_cluster)
+{
+    static bool _dummy = pegasus_client_factory::initialize(nullptr);
+
+    pegasus_client *client = pegasus_client_factory::get_client(remote_cluster.data(), app.data());
+    _client = static_cast<client::pegasus_client_impl *>(client);
+
+    auto ret = dsn::replication::get_duplication_cluster_id(remote_cluster.data());
+    dassert_replica(ret.is_ok(), // never possible, meta server disallows such remote_cluster.
```
Comment (hycdong) [inline src/server/pegasus_mutation_duplicator.cpp]: 这里的注释是说不会出现这种情况，因为meta server会做检查，不会出现根据cluster name得不到id的情况？这块的代码在哪里呢？我想再看一下

File: src/server/pegasus_mutation_duplicator.cpp [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+            dsn::apps::duplicate_request dreq;
+            dsn::from_blob_to_thrift(data, dreq);
+
+            auto timetag = static_cast<uint64_t>(dreq.timetag);
+            uint8_t from_cluster_id = extract_cluster_id_from_timetag(timetag);
+            if (from_cluster_id == _remote_cluster_id) {
+                // ignore this mutation to prevent infinite duplication loop.
+                continue;
+            }
+            if (!dsn::replication::is_cluster_id_configured(from_cluster_id)) {
```
Comment (hycdong) [inline src/server/pegasus_mutation_duplicator.cpp]: 这里是在判断什么情况？

File: src/server/pegasus_mutation_duplicator.cpp [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+    if (perr != PERR_OK || err != dsn::ERR_OK) {
+        _failed_shipping_ops->increment();
+
+        // randomly log the 1% of the failed duplicate rpc.
+        if (dsn::rand::next_double01() <= 0.01) {
+            derror_replica("duplicate_rpc failed: {} [code:{}, cluster_id:{}, timestamp:{}]",
+                           err == dsn::ERR_OK ? _client->get_error_string(perr) : err.to_string(),
+                           rpc.request().task_code,
+                           extract_cluster_id_from_timetag(rpc.request().timetag),
+                           extract_timestamp_from_timetag(rpc.request().timetag));
```
Comment (hycdong) [inline src/server/pegasus_mutation_duplicator.cpp]: 不把所有出错的duplicate请求打出来没问题嘛？

**Comment by hycdong** [comment]
本次pr热备是由pegasus_mutation_duplicator的duplicate函数开始，处理每个mutation_tuple_set，将每个mutation的dup rpc存储在_inflight map中，并依次调用send函数，pegasus server client的async_duplicate函数，pegasus client的duplicate函数，发送给远端的cluster，请问我的理解是否有问题？

**Comment by neverchanje** [comment]
> 本次pr涉及的复制流程是由pegasus_mutation_duplicator的duplicate函数开始，处理每个mutation_tuple_set，将每个mutation的dup rpc存储在_inflight map中，并依次调用send函数，pegasus server client的async_duplicate函数，pegasus client的duplicate函数，发送给远端的cluster，发送后的回调由on_duplicate_reply处理，请问我的理解是否有问题？

对

File: src/idl/rrdb.thrift [hunk 5]
```diff
@@ -251,6 +251,33 @@ struct scan_response
...
+
+    // The hash value calculated from the hash key, which is retrieved from `raw_message`.
+    // This field is to optimize when duplicating a DUPLICATE, the hash calculation can be skipped.
+    4: optional i64 hash
+
+    // Where the write is from. If the topology is A<->B<->C, a write to A will be then
+    // duplicated to B and C. For C, because. A and B is in the `from_clusters_set`, the write
+    // will not be duplicated to them, so that prevents infinite loop.
+    // The receiver of the write will
+    5: optional set<byte> from_clusters_set
```
Comment (levy5307) [inline src/idl/rrdb.thrift]: 这里加了之后，是不是替换掉了之前为了避免循环写引入的cluster id？

File: src/server/pegasus_mutation_duplicator.cpp [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+} // namespace dsn
+
+namespace pegasus {
+namespace server {
+
+using namespace dsn::literals::chrono_literals;
+
+/*extern*/ uint64_t get_hash_from_request(dsn::task_code tc, const dsn::blob &data)
+{
+    if (tc == dsn::apps::RPC_RRDB_RRDB_PUT) {
```
Comment (levy5307) [inline src/server/pegasus_mutation_duplicator.cpp]: 这里可以改成switch吗，感觉一堆if看着有点乱

Comment (neverchanje) [reply src/server/pegasus_mutation_duplicator.cpp]: 不行啊，RPC_RRDB_RRDB_PUT 看起来是常量，其实并不是常量，在 c++ 里，非 constant 不能用于 switch case.

File: src/server/pegasus_mutation_duplicator.cpp [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+        auto dreq = dsn::make_unique<dsn::apps::duplicate_request>();
+
+        // extract the rpc wrapped inside if this is a DUPLICATE rpc
+        if (rpc_code == dsn::apps::RPC_RRDB_RRDB_DUPLICATE) {
+            dsn::from_blob_to_thrift(raw_message, *dreq);
+            if (dreq->from_clusters_set.find(_remote_cluster_id) != dreq->from_clusters_set.end()) {
+                // ignore this mutation to prevent infinite duplication loop.
+                continue;
+            }
+            if (has_unknown_cluster_id(dreq->from_clusters_set)) {
```
Comment (levy5307) [inline src/server/pegasus_mutation_duplicator.cpp]: 什么情况下会收到unknown cluster的duplication?

File: src/server/pegasus_mutation_duplicator.cpp [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+        COUNTER_TYPE_RATE,
+        "the qps of failed DUPLICATE requests sent from this app");
+}
+
+void pegasus_mutation_duplicator::send(uint64_t hash, callback cb)
+{
+    duplicate_rpc rpc;
+    {
+        dsn::zauto_lock _(_lock);
+        rpc = _inflights[hash].front();
```
Comment (levy5307) [inline src/server/pegasus_mutation_duplicator.cpp]: 我理解这里是将相同的key批次发送出去，但是会不会出现不同的key有相同的hash呢？尤其对于RPC_RRDB_RRDB_PUT/RPC_RRDB_RRDB_REMOVE/RPC_RRDB_RRDB_MULTI_PUT/RPC_RRDB_RRDB_MULTI_REMOVE他们用的hash函数接口都不一样

File: src/idl/rrdb.thrift (hunk 5 above)
Comment (neverchanje) [reply src/idl/rrdb.thrift]: 是的，如注释所述

File: src/server/pegasus_mutation_duplicator.cpp [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+{
+    if (tc == dsn::apps::RPC_RRDB_RRDB_PUT) {
+        dsn::apps::update_request thrift_request;
+        dsn::from_blob_to_thrift(data, thrift_request);
+        return pegasus_key_hash(thrift_request.key);
+    }
+    if (tc == dsn::apps::RPC_RRDB_RRDB_REMOVE) {
+        dsn::blob raw_key;
+        dsn::from_blob_to_thrift(data, raw_key);
+        return pegasus_key_hash(raw_key);
```
Comment (levy5307) [inline src/server/pegasus_mutation_duplicator.cpp]: 这里为什么不同的操作要使用不同的hash函数？

File: src/idl/rrdb.thrift (hunk 5 above)
Comment (levy5307) [reply src/idl/rrdb.thrift]: 那文档也同步了吧

**Pull request merged**